### PR TITLE
Highlight features in text for FeatureUnion

### PR DESCRIPTION
**Pull request by lopuhin**
The approach is to highlight features for each of the vectorizers of the FeatureUnion, then render them and merge results.

TODO:

- [x] the bias term is not displayed for some reason
- [x] ~check (and test) support for unhashed features: something might be missing at the moment~ they don't work and I'd like to do it in a separate PR
- [x] tests for FeatureUnion, including cases of mixing text and non-text vectorizers

### CODE REVIEW

**Comment by kmike** [comment]
Hm, can we move span merging to explanations instead of having it in formatters?

**Comment by lopuhin** [comment]
> Hm, can we move span merging to explanations instead of having it in formatters?

Yes, that's what I originally did, but it had two drawbacks that lead me to move it to formatters:
- the code to merge highlightings is slightly more complex (but it's just 5 lines to re-calculate span starts and ends, so it's not a problem)
- right now, I use some html features for better display of merged highlitings - ``<br/>`` to separate lines, ``<b>`` to make vectorizer name bold



But yes, I see the point that it's better to have less code and logic in the formatters. Maybe it's possible to do most merging outside of them, and just have a list of (vectorizer name, document, spans) that are later passed to the formatters.

**Comment by kmike** [comment]
I think the tests could be the following: 

1) can we make it easier to create a formatter which takes JSON output and renders it on a client?
2) can we make it easier to highlight text if vectorizers are not from scikit-learn?

It looks like by moving code to formatters you helped (2), but not (1), while having this code directly in explain_.. functions will help (1), but not (2).

**Comment by lopuhin** [comment]
Right, I forgot about (2). I think it's not a huge problem, most code should be extractable into helpers.
I'll try to refactor the code to have more logic in the explanation part, and to make it generally less confusing :)

**Comment by kmike** [comment]
+1 to helpers; they can be easier to test as well

**Comment by lopuhin** [comment]
I think I'm finally more or less happy with how it works and how it is structured, still need to update unit tests after last refactoring, update docstrings and clean up stuff.

File: eli5/formatters/text_helpers.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def get_char_weights(doc_weighted_spans, preserve_density=None):
+    # type: (DocWeightedSpans, bool) -> np.ndarray
+    """ Return character weights for a text document with highlighted features.
+    If preserve_density is True, then color for longer fragments will be
+    less intensive than for shorter fragments, so that "sum" of intensities
+    will correspond to feature weight.
+    If preserve_density is None, then it's value is chosen depending on
+    analyzer kind: it is preserved for "char" and "char_wb" analyzers,
```
Comment (kmike) [inline eli5/formatters/text_helpers.py]: It'd be good for formatters not to be aware of scikit-learn. Maybe it is just a matter of documenting things differently.

**Comment by lopuhin** [comment]
@kmike this is finally ready for review :)

File: eli5/formatters/utils.py [hunk 2]
```diff
@@ -68,3 +68,8 @@ def _has_invisible_spaces(name):
     if isinstance(name, list):
         return any(_has_invisible_spaces(n['name']) for n in name)
     return name.startswith(' ') or name.endswith(' ')
+
+
+def max_or_0(it):
```
Comment (kmike) [inline eli5/formatters/utils.py]: I think eli5.utils is a better place for it; it is a general-purpose function not related to formatters directly. Could you please also add a test for it (a doctest is fine)?

File: tests/test_sklearn_vectorizers.py [hunk 3]
```diff
@@ -159,3 +161,74 @@ def test_explain_regression_hashing_vectorizer(newsgroups_train_binary):
...
+        {'url': 'http://c.com/blog',
+         'text': 'security'},
+        ]
+    ys = [1, 0, 0, 0, 1]
+    vec = FeatureUnion([
+        ('url', vec_cls(preprocessor=lambda x: x['url'],
+                        analyzer='char',
+                        ngram_range=(3, 3))),
+        ('text', vec_cls(preprocessor=lambda x: x['text'])),
+    ])
```
Comment (kmike) [inline tests/test_sklearn_vectorizers.py]: Is FeatureUnion for the same document supported (e.g. analyzer='char' CountVectorizer and analyzer='word' CountVectorizer at the same time)? This feature could be awesome for LIME.

File: eli5/formatters/utils.py (hunk 2 above)
Comment (lopuhin) [inline eli5/formatters/utils.py]: Yes, I agree, it's not specific to formatters at all. Moved in 9cba801, thanks!

File: tests/test_sklearn_vectorizers.py (hunk 3 above)
Comment (lopuhin) [inline tests/test_sklearn_vectorizers.py]: Yes, but they would be still shown as two separate "documents" (although the same) with features from each of the highlighted using different scales. Is this what you have in mind for LIME, or you would like to have just one "document" with features from both vectorizers applied?

Comment (kmike) [inline tests/test_sklearn_vectorizers.py]: I think showing it in a same document with the same scale (at least optionally) makes more sense, because it is easier to understand which parts of document are important overall. 

Yeah, this is what I had in mind for LIME - I was thinking about a default classifier to use, and 
a logistic regression trained on both char ngram and word unigram features looked like a baseline which should cover a lot of cases. I don't know how what it takes to implement this though :)

**Comment by kmike** [comment]
Thanks @lopuhin, yet another awesome feature from you!

File: tests/test_sklearn_vectorizers.py (hunk 3 above)
Comment (lopuhin) [inline tests/test_sklearn_vectorizers.py]: Yes, I see, thanks for the explanation! think it should be possible to modify an explanation that contains two separate documents into one by concatenating ``span`` fields of ``DocWeightedSpans``.

**Comment by lopuhin** [comment]
Thanks for review @kmike !
